File: server/resolvers/Artist.js
export default {
  albums: ({ id }, { limit = 50, offset = 0 }, { dataSources }) =>
    dataSources.spotifyAPI.getArtistAlbums(id, { limit, offset }),
  topTracks: ({ id }, { limit = 10 }, { dataSources }) =>
    dataSources.spotifyAPI
      .getTopTracksByArtist(id)
      .then(({ tracks }) => tracks.slice(0, limit)),
  relatedArtists: ({ id }, _args, { dataSources }) =>
    dataSources.spotifyAPI.getRelatedArtists(id).then(prop('artists'))
}

File: lib/spotify_web/resolvers/artist.ex
defmodule SpotifyWeb.Resolvers.Artist do
  alias SpotifyWeb.Resolvers.Helpers

  def albums(%{id: id}, args, %{context: %{authorization: authorization}}) do
    id
    |> Spotify.Client.albums_by_artist(args, [{"Authorization", authorization}])
    |> Helpers.handle_response()
  end

  def find(%{id: id}, %{context: %{authorization: authorization}}) do
    id
    |> Spotify.Client.find_artist([{"Authorization", authorization}])
    |> Helpers.handle_response()
  end

  def related_artists(%{id: id}, _, res) do
    id
    |> Spotify.Client.related_artists_for_artist(Helpers.prepare_headers(res))
    |> Helpers.handle_response(fn %{artists: artists} -> artists end)
  end

  def top_tracks(%{id: id}, %{limit: limit}, %{
        context: %{authorization: authorization}
      }) do
    id
    |> Spotify.Client.top_tracks_by_artist([{"Authorization", authorization}])
    |> Helpers.handle_response(fn %{tracks: tracks} ->
      Enum.take(tracks, limit)
    end)
  end
end
